kratos/utilities/assign_mpcs_to_neighbours_utility.cpp

```diff
@@ -0,0 +1,376 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main author:     Sebastian Ares de Parga Regalado
+//
+
+// System includes
+#include <string>
+#include <iostream>
+
+// External includes
+#include "concurrentqueue/concurrentqueue.h"
+
+// Project includes
+#include "utilities/assign_mpcs_to_neighbours_utility.h"
+
+namespace Kratos 
+{
+///@addtogroup Kratos Core
+///@{
+
+///@name Kratos Classes
+///@{
+
+/**
+ * @class AssignMPCsToNeighboursUtility
+ * @ingroup Kratos Core
+ * @brief Assing MPCs to Neighbouring Nodes
+ * @details This class provides a method to search for neighbouring nodes of one node
+ * within a given radius and assign a multipoint constraint (MPC) using radial basis functions.
+ * @author Sebastian Ares de Parga Regalado
+ */
+
+/// Default constructor.
+AssignMPCsToNeighboursUtility::AssignMPCsToNeighboursUtility(NodesContainerType& rStructureNodes){
+    KRATOS_TRY;
+    NodesContainerType::ContainerType& nodes_model_part = rStructureNodes.GetContainer();
+    mpBins = Kratos::make_unique<NodeBinsType>(nodes_model_part.begin(), nodes_model_part.end());
+    mMaxNumberOfNodes = rStructureNodes.size();
+    KRATOS_CATCH(""); 
+} 
+
+// Search for the neighbouring nodes (in rStructureNodes) of each rNode on a given array of rNodes
+void AssignMPCsToNeighboursUtility::SearchNodesInRadiusForNodes(
+    NodesContainerType const& rNodes,
+    const double Radius,
+    VectorResultNodesContainerType& rResults,
+    const double MinNumOfNeighNodes)
+{
+    KRATOS_TRY;
+    NodesContainerType::ContainerType& nodes_array     = const_cast<NodesContainerType::ContainerType&>(rNodes.GetContainer());
+
+    rResults.resize(nodes_array.size());
```

Comment:
This means you're allocating much memory than needed. I'd do a `shrink_to_fit` at the end of this function to avoid so.

Reply:
I am allocating what I need. Nonetheless, I could updated it to this:
```cpp
    // Resize rResults vector
    rResults.resize(r_nodes_array.size());

    // Declare a counter variable outside the loop as std::atomic<int>
    std::atomic<int> i(0);

    block_for_each(r_nodes_array, [&](const Node<3>::Pointer& pNode)
    {
        double local_radius = Radius;

        IndexType it = i.fetch_add(1); // Atomically increment the counter and get the previous value

        while (rResults[it].size()<MinNumOfNeighNodes){
            rResults[it].clear();
            SearchNodesInRadiusForNode(pNode, local_radius, rResults[it]);
            local_radius += Radius;
        }
    });

    // Optimize memory usage
    rResults.shrink_to_fit();

Reply:
No. You are allocating much more than what you need since you're resizing the vector to the number of nodes, which would only be the case if all the nodes are found. Indeed, the proper usage would be
```c++
rResults.reserve(r_nodes_array.size());

// ... your stuff in here ...

rResults.shrink_to_fit();
```

Reply:
I am changing the names of the functions to make them more self-explanatory. In this function, we retrieve the degrees of freedom (DOFs) and coordinates of the cloud of nodes. Therefore, we already know the size.